src/protocol/attribution/apply_attribution_window.rs

```diff
@@ -0,0 +1,351 @@
+///
+/// # Errors
+/// Fails if sub-protocols fails.
+#[allow(dead_code)]
+async fn apply_attribution_window<F, C, T>(
+    ctx: C,
+    input: &[MCApplyAttributionWindowInputRow<F, T>],
+    attribution_window_seconds: u32,
+) -> Result<Vec<MCApplyAttributionWindowOutputRow<F, T>>, Error>
+where
+    F: Field,
+    C: Context<F, Share = T>,
+    T: Arithmetic<F>,
+{
+    let t_deltas = prefix_sum_time_deltas(&ctx, input).await?;
+
+    let trigger_values =
+        zero_out_expired_trigger_values(&ctx, input, &t_deltas, attribution_window_seconds).await?;
+
+    Ok(input
+        .iter()
+        .zip(trigger_values)
+        .map(|(x, value)| {
+            MCApplyAttributionWindowOutputRow::new(
+                x.is_trigger_report.clone(),
+                x.helper_bit.clone(),
+                x.breakdown_key.clone(),
+                value,
+            )
+        })
+        .collect::<Vec<_>>())
+}
+
+/// Computes time deltas from each trigger event to its nearest matching source event.
+///
+/// # Errors
+/// Fails if the multiplication fails.
+async fn prefix_sum_time_deltas<F, C, T>(
+    ctx: &C,
+    input: &[MCApplyAttributionWindowInputRow<F, T>],
+) -> Result<Vec<T>, Error>
+where
+    F: Field,
+    C: Context<F, Share = T>,
+    T: Arithmetic<F>,
+{
+    let num_rows = input.len();
+
+    // Pre-compute `is_trigger_bit * helper_bit'.
+    // Later in the prefix-sum loop, this vector is updated in each iteration to help
+    // accumulate timedeltas and determine when to stop accumulating.
+    let stop_bit_context = ctx
+        .narrow(&Step::IsTriggerBitTimesHelperBit)
+        .set_total_records(num_rows - 1);
+    let mut stop_bits = Some(T::ZERO)
+        .into_iter()
+        .chain(
+            try_join_all(input.iter().skip(1).enumerate().map(|(i, x)| {
+                let c = stop_bit_context.clone();
+                let record_id = RecordId::from(i);
+                async move {
+                    c.multiply(record_id, &x.is_trigger_report, &x.helper_bit)
+                        .await
+                }
+            }))
+            .await?,
+        )
+        .collect::<Vec<_>>();
+
+    // First, create a vector of timedeltas. This vector contains non-zero values only for
+    // rows with `stop_bit` = 1, meaning that the row is a trigger event, and has the same
+    // match key as the event one above.
+    let t_delta_context = ctx
+        .narrow(&Step::InitializeTimeDelta)
+        .set_total_records(num_rows - 1);
+    let mut t_delta = Some(T::ZERO)
+        .into_iter()
+        .chain(
+            try_join_all(
+                zip(input.iter(), input.iter().skip(1))
+                    .zip(stop_bits.iter().skip(1))
+                    .enumerate()
+                    .map(|(i, ((prev, curr), b))| {
+                        let c = t_delta_context.clone();
+                        let record_id = RecordId::from(i);
+                        let delta = curr.timestamp.clone() - &prev.timestamp;
```

Comment:
this clone is annoying - one day I tried to generalize `Add` for `ArithmeticRefOps` to work for `&T` and it was a spectacular waste of 3 hours. If someone knows how to do it, please let me know